pkg/cassandra/config.go

```diff
@@ -75,32 +75,36 @@ func (c config) MarshalJSON() ([]byte, error) {
 	return json.Marshal(&config)
 }
 
-func newConfig(apiConfig *api.CassandraConfig, cassandraVersion string) config {
-	config := config{cassandraVersion: cassandraVersion}
+func newConfig(apiConfig *api.CassandraConfig, cassandraVersion string) (config, error) {
```

Comment:
These types and name refactoring look odd. In this place the config name shadowing has been modified, but in other places in the same file they haven't been.

Reply:
I can't have `config` shadowed because I need to use it again [here](https://github.com/Miles-Garnsey/k8ssandra-operator/blob/1dfe518f9add98a327825f7aee8cbb668aa54e17/pkg/cassandra/config.go#L82).

What are you asking me to change here?

Reply:
NB: I actually can't see another instance where the internal `config` struct is shadowed by a variable named `config`.

Reply:
I mean you should **NOT** shadow it. You removed shadowing in one place, but not in others (which you still modified). Consistency.

Reply:
I'm still not grokking what you're asking for I'm sorry. Are you suggesting that I should replace `config` with `cfg` for local variables everywhere in this file? The only functions I've modified are `newConfig` and `CreateJsonConfig`. The latter already had a variable called `cfg` so if you're nitpicking with me about consistency, the code is no more inconsistent now than it was before.

The other functions (except `MarshalJSON`) don't use the `config` struct, so the shadowing doesn't matter for them. I do not believe that shadowing must be avoided in all cases.

The real problem here is that `config{}` is a bad name. It is up there with anti-patterns like `Xmanager` and `a`. In fact, I don't understand what the purpose of the internal `config{}` is. It seems to be designed purely as a place to attach `MarshalJSON()`, but that function isn't called anywhere.

But this isn't my code, and I'm not going to go at it with a hacksaw. Closing this item.

Reply:
>It seems to be designed purely as a place to attach `MarshalJSON()`, but that function isn't called anywhere.

The `config` struct, while it may suffer from a major naming anti-pattern up there with the likes of `Xmanager` and `a` :smile, actually provides some key logic. If you take a look at `config_test.go` you will find tests that exercise and verify this functionality. It's a good starting point to understanding the code involved.


The `config` struct is an adapter to provide custom marshaling. The `Marshaler` interface is defined in the `json` package. The `json.Marshal` function traverses the value recursively. If it is encounters an object that implements `Marshaler` it calls its `MarshalJSON` function.

If you look at the `MarshalJSON` function you will notice that it does several things:

* If the version is Cassandra 4, remove properties that are only relevant to Cassandra 3.x 
* If `NumTokens` is `nil` set a default value based on Cassandra version
* Make sure appropriate `jvm-options` or `jvm-server-options` is used based on Cassandra version

>But this isn't my code, and I'm not going to go at it with a hacksaw

It's not your code. It's not my code. It's our code. And I certainly agree that you shouldn't go at it with a hacksaw, at least not until you have a better understanding of it 😉